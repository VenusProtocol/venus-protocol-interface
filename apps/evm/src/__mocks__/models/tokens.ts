import bnbLogo from 'packages/tokens/img/bnb.svg';
import busdLogo from 'packages/tokens/img/busd.svg';
import ethLogo from 'packages/tokens/img/eth.svg';
import lisUsdLogo from 'packages/tokens/img/lisUSD.png';
import lunaLogo from 'packages/tokens/img/luna.svg';
import usdcLogo from 'packages/tokens/img/usdc.svg';
import usdtLogo from 'packages/tokens/img/usdt.svg';
import ustLogo from 'packages/tokens/img/ust.svg';
import vaiLogo from 'packages/tokens/img/vai.svg';
import vrtLogo from 'packages/tokens/img/vrt.svg';
import wbnbLogo from 'packages/tokens/img/wbnb.svg';
import xvsLogo from 'packages/tokens/img/xvs.svg';
import { Token } from 'types';

export const xvs: Token = {
  address: '0xB9e0E753630434d7863528cc73CB7AC638a7c8ff',
  decimals: 18,
  symbol: 'XVS',
  asset: xvsLogo,
};

export const vai: Token = {
  address: '0x5fFbE5302BadED40941A403228E6AD03f93752d9',
  decimals: 18,
  symbol: 'VAI',
  asset: vaiLogo,
};

export const vrt: Token = {
  address: '0x331F639B4F3CF6E391CD244e0b5027C5968Ec325',
  decimals: 18,
  symbol: 'VRT',
  asset: vrtLogo,
};

export const bnb: Token = {
  address: '0x0000000000000000000000000000000000000000',
  isNative: true,
  decimals: 18,
  symbol: 'BNB',
  asset: bnbLogo,
};

export const usdc: Token = {
  address: '0x16227D60f7a0e586C66B005219dfc887D13C9531',
  decimals: 6,
  symbol: 'USDC',
  asset: usdcLogo,
};

export const usdt: Token = {
  address: '0xA11c8D9DC9b66E209Ef60F0C8D969D3CD988782c',
  decimals: 6,
  symbol: 'USDT',
  asset: usdtLogo,
};

export const busd: Token = {
  address: '0x8301F2213c0eeD49a7E28Ae4c3e91722919B8B47',
  decimals: 18,
  symbol: 'BUSD',
  asset: busdLogo,
};

export const ust: Token = {
  address: '0x5A79efD958432E72211ee73D5DDFa9bc8f248b5F',
  decimals: 18,
  symbol: 'UST',
  asset: ustLogo,
};

export const luna: Token = {
  address: '0xf36160EC62E3B191EA375dadfe465E8Fa1F8CabB',
  decimals: 6,
  symbol: 'LUNA',
  asset: lunaLogo,
};

export const wbnb: Token = {
  symbol: 'WBNB',
  decimals: 18,
  address: '0xae13d989daC2f0dEbFf460aC112a837C89BAa7cd',
  asset: wbnbLogo,
};

export const eth: Token = {
  address: '0x98f7A83361F7Ac8765CcEBAB1425da6b341958a7',
  decimals: 18,
  symbol: 'ETH',
  asset: ethLogo,
};

export const lisUsd: Token = {
  address: '0xe73774DfCD551BF75650772dC2cC56a2B6323453',
  decimals: 18,
  symbol: 'lisUSD',
  asset: lisUsdLogo,
};

export default [xvs, bnb, usdc, usdt, busd, ust, luna, vai, wbnb, eth, vrt, lisUsd];
