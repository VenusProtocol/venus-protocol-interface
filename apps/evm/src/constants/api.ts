import type { Network } from 'types';

export const apiUrls: Record<Network, string> = {
  testnet: 'https://testnetapi.venus.io',
  mainnet: 'https://api.venus.io',
  'mainnet-preview': 'https://api-preview.venus.io',
};
