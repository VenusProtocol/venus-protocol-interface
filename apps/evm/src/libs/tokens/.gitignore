# Automatically generated PancakeSwap token list
/generated