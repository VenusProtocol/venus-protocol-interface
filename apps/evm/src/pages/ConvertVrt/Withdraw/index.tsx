/** @jsxImportSource @emotion/react */
import { Typography } from '@mui/material';
import BigNumber from 'bignumber.js';

import { PrimaryButton } from 'components';
import { ConnectWallet } from 'containers/ConnectWallet';
import useConvertMantissaToReadableTokenString from 'hooks/useConvertMantissaToReadableTokenString';
import { displayMutationError } from 'packages/errors';
import { useLunaUstWarning } from 'packages/lunaUstWarning';
import { useGetToken } from 'packages/tokens';
import { useTranslation } from 'packages/translations';

import { useStyles } from '../styles';

export interface WithdrawProps {
  xvsWithdrawableAmount: BigNumber | undefined;
  withdrawXvs: () => Promise<unknown>;
  withdrawXvsLoading: boolean;
}

const Withdraw: React.FC<WithdrawProps> = ({
  xvsWithdrawableAmount,
  withdrawXvs,
  withdrawXvsLoading,
}) => {
  const { t } = useTranslation();
  const styles = useStyles();
  const xvs = useGetToken({
    symbol: 'XVS',
  });

  const { userHasLunaOrUstCollateralEnabled, openLunaUstWarningModal } = useLunaUstWarning();

  const readableXvsAvailable = useConvertMantissaToReadableTokenString({
    value: xvsWithdrawableAmount,
    token: xvs,
  });

  const onSubmit = async () => {
    // Block action is user has LUNA or UST enabled as collateral
    if (userHasLunaOrUstCollateralEnabled) {
      openLunaUstWarningModal();
      return;
    }

    try {
      await withdrawXvs();
    } catch (error) {
      displayMutationError({ error });
    }
  };

  return (
    <div css={styles.root}>
      <ConnectWallet message={t('convertVrt.connectWalletToWithdrawXvs')}>
        <section css={styles.title}>
          <Typography variant="h3">{readableXvsAvailable}</Typography>
          <Typography variant="small2">{t('convertVrt.withdrawableAmount')}</Typography>
        </section>
        <PrimaryButton
          disabled={
            !xvsWithdrawableAmount ||
            xvsWithdrawableAmount.isZero() ||
            xvsWithdrawableAmount.isNaN()
          }
          className="w-full"
          onClick={onSubmit}
          loading={withdrawXvsLoading}
        >
          {t('convertVrt.withdrawXvs')}
        </PrimaryButton>
      </ConnectWallet>
    </div>
  );
};

export default Withdraw;
