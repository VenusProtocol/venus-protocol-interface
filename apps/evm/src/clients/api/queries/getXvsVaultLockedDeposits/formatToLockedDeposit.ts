import BigNumber from 'bignumber.js';

import { XvsVault } from 'libs/contracts';
import { LockedDeposit } from 'types';

const formatToLockedDeposit = ([amount, lockedUntil]: Awaited<
  ReturnType<XvsVault['getWithdrawalRequests']>
>[number]): LockedDeposit => {
  // lockedUntil is a timestamp expressed in seconds, so we convert it to milliseconds
  const lockedUntilMs = +lockedUntil * 1000;
  const unlockedAt = new Date(lockedUntilMs);

  return {
    amountMantissa: new BigNumber(amount.toString()),
    unlockedAt,
  };
};

export default formatToLockedDeposit;
