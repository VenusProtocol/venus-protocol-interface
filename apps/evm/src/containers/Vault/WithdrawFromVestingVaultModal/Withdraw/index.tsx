/** @jsxImportSource @emotion/react */
import BigNumber from 'bignumber.js';
import { isBefore } from 'date-fns/isBefore';
import { useMemo } from 'react';

import { useExecuteWithdrawalFromXvsVault, useGetXvsVaultLockedDeposits } from 'clients/api';
import { LabeledInlineContent, PrimaryButton, Spinner } from 'components';
import { ConnectWallet } from 'containers/ConnectWallet';
import useConvertMantissaToReadableTokenString from 'hooks/useConvertMantissaToReadableTokenString';
import { useGetToken } from 'libs/tokens';
import { useTranslation } from 'libs/translations';
import { useAccountAddress } from 'libs/wallet';
import type { Token } from 'types';

import { NULL_ADDRESS } from 'constants/address';
import { SwitchChain } from 'containers/SwitchChain';
import TEST_IDS from './testIds';

export interface WithdrawUiProps {
  stakedToken: Token;
  isInitialLoading: boolean;
  onSubmitSuccess: () => void;
  onSubmit: () => Promise<unknown>;
  isSubmitting: boolean;
  withdrawableMantissa?: BigNumber;
}

const WithdrawUi: React.FC<WithdrawUiProps> = ({
  stakedToken,
  isInitialLoading,
  onSubmit,
  onSubmitSuccess,
  isSubmitting,
  withdrawableMantissa,
}) => {
  const { t } = useTranslation();

  const handleSubmit = async () => {
    await onSubmit();

    onSubmitSuccess();
  };

  const readableWithdrawableTokens = useConvertMantissaToReadableTokenString({
    value: withdrawableMantissa,
    token: stakedToken,
  });

  const hasWithdrawableTokens = !!withdrawableMantissa && withdrawableMantissa.isGreaterThan(0);

  let submitDom = (
    <PrimaryButton
      type="submit"
      onClick={handleSubmit}
      loading={isSubmitting}
      disabled={!hasWithdrawableTokens}
      className="w-full"
    >
      {t('withdrawFromVestingVaultModalModal.withdrawTab.submitButton')}
    </PrimaryButton>
  );

  if (hasWithdrawableTokens) {
    submitDom = <SwitchChain>{submitDom}</SwitchChain>;
  }

  return (
    <>
      {isInitialLoading || !withdrawableMantissa ? (
        <Spinner />
      ) : (
        <div className="space-y-6">
          <LabeledInlineContent
            iconSrc={stakedToken}
            data-testid={TEST_IDS.availableTokens}
            label={t('withdrawFromVestingVaultModalModal.withdrawTab.availableTokens', {
              tokenSymbol: stakedToken.symbol,
            })}
          >
            {readableWithdrawableTokens}
          </LabeledInlineContent>

          {submitDom}
        </div>
      )}
    </>
  );
};

export interface WithdrawProps {
  stakedToken: Token;
  poolIndex: number;
  handleClose: () => void;
}

const Withdraw: React.FC<WithdrawProps> = ({ stakedToken, poolIndex, handleClose }) => {
  const { t } = useTranslation();
  const { accountAddress } = useAccountAddress();

  const xvs = useGetToken({
    symbol: 'XVS',
  });

  const {
    data: xvsVaultUserLockedDepositsData = {
      lockedDeposits: [],
    },
    isLoading: isGetXvsVaultUserLockedDepositsLoading,
  } = useGetXvsVaultLockedDeposits(
    {
      poolIndex,
      rewardTokenAddress: xvs!.address,
      accountAddress: accountAddress || NULL_ADDRESS,
    },
    {
      placeholderData: {
        lockedDeposits: [],
      },
      enabled: !!accountAddress,
    },
  );

  const withdrawableMantissa = useMemo(() => {
    const now = new Date();

    return xvsVaultUserLockedDepositsData.lockedDeposits.reduce(
      (acc, xvsVaultUserLockedDeposit) =>
        isBefore(xvsVaultUserLockedDeposit.unlockedAt, now)
          ? acc.plus(xvsVaultUserLockedDeposit.amountMantissa)
          : acc,
      new BigNumber(0),
    );
  }, [xvsVaultUserLockedDepositsData]);

  const {
    mutateAsync: executeWithdrawalFromXvsVault,
    isPending: isExecutingWithdrawalFromXvsVault,
  } = useExecuteWithdrawalFromXvsVault({
    stakedToken,
  });

  const handleSubmit = () =>
    executeWithdrawalFromXvsVault({
      poolIndex,
      rewardTokenAddress: xvs!.address,
    });

  return (
    <ConnectWallet
      analyticVariant="vesting_vault_withdraw_modal"
      message={t(
        'withdrawFromVestingVaultModalModal.withdrawTab.approveToken.connectWalletMessage',
      )}
    >
      <WithdrawUi
        stakedToken={stakedToken}
        isInitialLoading={isGetXvsVaultUserLockedDepositsLoading}
        isSubmitting={isExecutingWithdrawalFromXvsVault}
        withdrawableMantissa={withdrawableMantissa}
        onSubmit={handleSubmit}
        onSubmitSuccess={handleClose}
      />
    </ConnectWallet>
  );
};

export default Withdraw;
