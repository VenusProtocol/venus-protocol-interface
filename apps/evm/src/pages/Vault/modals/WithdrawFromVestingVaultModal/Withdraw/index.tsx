/** @jsxImportSource @emotion/react */
import BigNumber from 'bignumber.js';
import { isBefore } from 'date-fns/isBefore';
import { useMemo } from 'react';

import { useExecuteWithdrawalFromXvsVault, useGetXvsVaultLockedDeposits } from 'clients/api';
import { LabeledInlineContent, PrimaryButton, Spinner } from 'components';
import { ConnectWallet } from 'containers/ConnectWallet';
import useConvertMantissaToReadableTokenString from 'hooks/useConvertMantissaToReadableTokenString';
import { useGetToken } from 'libs/tokens';
import { useTranslation } from 'libs/translations';
import { useAccountAddress } from 'libs/wallet';
import type { Token } from 'types';

import { useStyles } from './styles';
import TEST_IDS from './testIds';

export interface WithdrawUiProps {
  stakedToken: Token;
  isInitialLoading: boolean;
  onSubmitSuccess: () => void;
  onSubmit: () => Promise<unknown>;
  isSubmitting: boolean;
  withdrawableMantissa?: BigNumber;
}

const WithdrawUi: React.FC<WithdrawUiProps> = ({
  stakedToken,
  isInitialLoading,
  onSubmit,
  onSubmitSuccess,
  isSubmitting,
  withdrawableMantissa,
}) => {
  const { t } = useTranslation();
  const styles = useStyles();

  const handleSubmit = async () => {
    await onSubmit();

    onSubmitSuccess();
  };

  const readableWithdrawableTokens = useConvertMantissaToReadableTokenString({
    value: withdrawableMantissa,
    token: stakedToken,
  });

  return (
    <>
      {isInitialLoading || !withdrawableMantissa ? (
        <Spinner />
      ) : (
        <>
          <LabeledInlineContent
            css={styles.content}
            iconSrc={stakedToken}
            data-testid={TEST_IDS.availableTokens}
            label={t('withdrawFromVestingVaultModalModal.withdrawTab.availableTokens', {
              tokenSymbol: stakedToken.symbol,
            })}
          >
            {readableWithdrawableTokens}
          </LabeledInlineContent>

          <PrimaryButton
            type="submit"
            onClick={handleSubmit}
            loading={isSubmitting}
            disabled={withdrawableMantissa.isEqualTo(0)}
            className="w-full"
          >
            {t('withdrawFromVestingVaultModalModal.withdrawTab.submitButton')}
          </PrimaryButton>
        </>
      )}
    </>
  );
};

export interface WithdrawProps {
  stakedToken: Token;
  poolIndex: number;
  handleClose: () => void;
}

const Withdraw: React.FC<WithdrawProps> = ({ stakedToken, poolIndex, handleClose }) => {
  const { t } = useTranslation();
  const { accountAddress } = useAccountAddress();

  const xvs = useGetToken({
    symbol: 'XVS',
  });

  const {
    data: xvsVaultUserLockedDepositsData = {
      lockedDeposits: [],
    },
    isLoading: isGetXvsVaultUserLockedDepositsLoading,
  } = useGetXvsVaultLockedDeposits(
    {
      poolIndex,
      rewardTokenAddress: xvs!.address,
      accountAddress: accountAddress || '',
    },
    {
      placeholderData: {
        lockedDeposits: [],
      },
      enabled: !!accountAddress,
    },
  );

  const withdrawableMantissa = useMemo(() => {
    const now = new Date();

    return xvsVaultUserLockedDepositsData.lockedDeposits.reduce(
      (acc, xvsVaultUserLockedDeposit) =>
        isBefore(xvsVaultUserLockedDeposit.unlockedAt, now)
          ? acc.plus(xvsVaultUserLockedDeposit.amountMantissa)
          : acc,
      new BigNumber(0),
    );
  }, [xvsVaultUserLockedDepositsData]);

  const {
    mutateAsync: executeWithdrawalFromXvsVault,
    isPending: isExecutingWithdrawalFromXvsVault,
  } = useExecuteWithdrawalFromXvsVault({
    stakedToken,
  });

  const handleSubmit = () =>
    executeWithdrawalFromXvsVault({
      poolIndex,
      rewardTokenAddress: xvs!.address,
    });

  return (
    <ConnectWallet
      message={t(
        'withdrawFromVestingVaultModalModal.withdrawTab.approveToken.connectWalletMessage',
      )}
    >
      <WithdrawUi
        stakedToken={stakedToken}
        isInitialLoading={isGetXvsVaultUserLockedDepositsLoading}
        isSubmitting={isExecutingWithdrawalFromXvsVault}
        withdrawableMantissa={withdrawableMantissa}
        onSubmit={handleSubmit}
        onSubmitSuccess={handleClose}
      />
    </ConnectWallet>
  );
};

export default Withdraw;
