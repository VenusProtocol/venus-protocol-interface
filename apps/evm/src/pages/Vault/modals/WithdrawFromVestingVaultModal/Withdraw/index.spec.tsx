import { fireEvent, waitFor } from '@testing-library/react';
import noop from 'noop-ts';
import Vi from 'vitest';

import xvsVaultResponses from '__mocks__/contracts/xvsVault';
import fakeAddress from '__mocks__/models/address';
import { vai, xvs } from '__mocks__/models/tokens';
import { renderComponent } from 'testUtils/render';

import { executeWithdrawalFromXvsVault, getXvsVaultLockedDeposits } from 'clients/api';
import formatToLockedDeposit from 'clients/api/queries/getXvsVaultLockedDeposits/formatToLockedDeposit';
import { en } from 'libs/translations';

import Withdraw from '.';
import TEST_IDS from './testIds';

const fakePoolIndex = 6;
const fakeStakedToken = vai;

describe('pages/Vault/modals/WithdrawFromVestingVaultModal/Withdraw', () => {
  beforeEach(() => {
    vi.useFakeTimers().setSystemTime(new Date(1656603774626));
    (getXvsVaultLockedDeposits as Vi.Mock).mockImplementation(() => ({
      lockedDeposits: xvsVaultResponses.getWithdrawalRequests.map(formatToLockedDeposit),
    }));
  });

  it('renders without crashing', async () => {
    const { getByText } = renderComponent(
      <Withdraw poolIndex={fakePoolIndex} stakedToken={fakeStakedToken} handleClose={noop} />,
      {
        accountAddress: fakeAddress,
      },
    );

    await waitFor(() => getByText(en.withdrawFromVestingVaultModalModal.withdrawTab.submitButton));
  });

  it('fetches available tokens amount and displays it correctly', async () => {
    const { getByTestId } = renderComponent(
      <Withdraw poolIndex={fakePoolIndex} stakedToken={fakeStakedToken} handleClose={noop} />,
      {
        accountAddress: fakeAddress,
      },
    );

    await waitFor(() => getByTestId(TEST_IDS.availableTokens));

    expect(getByTestId(TEST_IDS.availableTokens).textContent).toMatchSnapshot();
  });

  it('disables submit button when there is no tokens available', async () => {
    (getXvsVaultLockedDeposits as Vi.Mock).mockImplementation(() => ({
      lockedDeposits: [],
    }));

    const { getByTestId, getByText } = renderComponent(
      <Withdraw poolIndex={fakePoolIndex} stakedToken={fakeStakedToken} handleClose={noop} />,
      {
        accountAddress: fakeAddress,
      },
    );

    await waitFor(() => getByTestId(TEST_IDS.availableTokens));

    const submitButton = getByText(
      en.withdrawFromVestingVaultModalModal.withdrawTab.submitButton,
    ).closest('button') as HTMLButtonElement;
    expect(submitButton).toBeDisabled();
  });

  it('lets user withdraw their available tokens and calls handleClose callback on success', async () => {
    const handleCloseMock = vi.fn();

    const { getByTestId, getByText } = renderComponent(
      <Withdraw
        poolIndex={fakePoolIndex}
        stakedToken={fakeStakedToken}
        handleClose={handleCloseMock}
      />,
      {
        accountAddress: fakeAddress,
      },
    );

    await waitFor(() => getByTestId(TEST_IDS.availableTokens));

    const submitButton = getByText(
      en.withdrawFromVestingVaultModalModal.withdrawTab.submitButton,
    ).closest('button') as HTMLButtonElement;

    // Click on submit button
    fireEvent.click(submitButton);

    await waitFor(() => expect(executeWithdrawalFromXvsVault).toHaveBeenCalledTimes(1));
    expect(executeWithdrawalFromXvsVault).toHaveBeenCalledWith({
      poolIndex: fakePoolIndex,
      rewardTokenAddress: xvs.address,
    });

    await waitFor(() => expect(handleCloseMock).toHaveBeenCalledTimes(1));
  });
});
