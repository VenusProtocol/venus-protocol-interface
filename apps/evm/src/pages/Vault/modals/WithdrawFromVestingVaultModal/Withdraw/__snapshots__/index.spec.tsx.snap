// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pages/Vault/modals/WithdrawFromVestingVaultModal/Withdraw > fetches available tokens amount and displays it correctly 1`] = `"Available VAI3.00 VAI"`;
